feat(image_projection_based_fusion): add filtering-based roi-detected-object fusion (#2178)
autowarefoundation/autoware.universe: perception/image_projection_based_fusion/docs/roi-detected-object-fusion.md

## Inner-workings / Algorithms

> The algorithm is shown below and see `Parameters` for the meaning of each parameter :

Reviewer: In what follows, we describe the algorithm utilized by `roi_detected_object_fusion` (the meaning of each parameter can be found in the `Parameters` section):

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

> 1. If `existence_probability` of detected objects is greater than the threshold, they passthrough without doing anything.

Reviewer: 1. If the `existence_probability` of a detected object is greater than the threshold, it is accepted without any further processing.

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

> 2. The rest of detected objects are projected onto image planes, and if ROIs are overlapped with them, they are output as fused objects. Intersection over Union (IoU) is used to determine if there are overlaps between them.

Reviewer: 2. The remaining detected objects are projected onto image planes, and if the resulting ROIs overlap with the ones from the 2D detector, they are published as fused objects. The Intersection over Union (IoU) is used to determine if there are overlaps between the detections from `input` and the ROIs from `input/rois`.

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

> The `DetectedObject` has three shape and the polygon vertices of a object are as below:

Reviewer: The wording makes it very difficult to read/understand.

The `DetectedObject` has three possible `shape` choices/implementations, where the polygon's vertices for each case are defined as follows:

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

| ------------------------- | ----------------------------------------------------- | -------------------------- |
| `output`                  | `autoware_auto_perception_msgs::msg::DetectedObjects` | detected objects           |
| `~/debug/image_raw[0-7]`  | `sensor_msgs::msg::Image`                             | images for visualization,  |
| `~/debug/fused_objects`   | `autoware_auto_perception_msgs::msg::DetectedObjects` | fused detected objects     |
| `~/debug/ignored_objects` | `autoware_auto_perception_msgs::msg::DetectedObjects` | not fused detected objects |

## Parameters

### Core Parameters

| Name                                            | Type   | Description                                                                                                  |
| ----------------------------------------------- | ------ | ------------------------------------------------------------------------------------------------------------ |
| `rois_number`                                   | int    | the number of input rois                                                                                     |

> | `debug_mode`                                    | bool   | If `true`, subscribe and publish images for visualization.                                                   |

Reviewer: If set to `true`, the node subscribes to the image topic and publishes an image with debug drawings.

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

> | `min_iou_threshold`                             | double | If the iou between detected objects and rois is greater than threshold, the objects are classified as fused. |

Reviewer: greater than `min_iou_threshold`

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

> | `passthrough_lower_bound_probability_threshold` | double | If `existence_probability` of detected objects is greater than the threshold, they passthrough as outputs    |

Reviewer: If the `existence_probability` of a detected object is greater than the threshold, it is published in `output`
*Note pass-through is a noun. In this case, it would be "they pass through"

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

| `use_roi_probability`                           | float  | use `existence_probability` of ROIs to match with detected objects                                           |

> | `roi_probability_threshold`                     | double | If `existence_probability` of ROIs is greater than the threshold, matched detected objects are output        |

Reviewer: output is a noun, so it can not be used in this manner.
"outputted" also sounds weird, so it may be better to phrase it like: "matched detected objects are published in `output`"

Author: updated at [886ce8a](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/886ce8a894c29e8fd0df34f56bc743980583a86b).

### Output

| Name                    | Type                                                  | Description                                       |
| ----------------------- | ----------------------------------------------------- | ------------------------------------------------- |
| `output`                | `autoware_auto_perception_msgs::msg::DetectedObjects` | detected objects                                  |

> | `output/image_rawID`    | `sensor_msgs::msg::Image`                             | images for visualization, `ID` is between 0 and 7 |

Reviewer 2: ```suggestion
| `~/output/debug/image_raw_[0~7]`    | `sensor_msgs::msg::Image`                             | images for visualization, `ID` is between 0 and 7 |
```

Author: fixed at [9c0ba6f](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/9c0ba6f356b0b4be884cde1ca8e0f2fdc7a8a55a).
I deleted `output`.

> | `debug/fused_objects`   | `autoware_auto_perception_msgs::msg::DetectedObjects` | fused detected objects                            |
> | `debug/ignored_objects` | `autoware_auto_perception_msgs::msg::DetectedObjects` | not fused detected objects                        |

Reviewer 2: ```suggestion
| `~/output/debug/fused_objects`   | `autoware_auto_perception_msgs::msg::DetectedObjects` | fused detected objects                            |
| `~/output/debug/ignored_objects` | `autoware_auto_perception_msgs::msg::DetectedObjects` | not fused detected objects                        |
```

Author: fixed at [9c0ba6f](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/9c0ba6f356b0b4be884cde1ca8e0f2fdc7a8a55a). I deleted `output`.

## Parameters

### Core Parameters

| Name                                            | Type   | Description                                                                                               |
| ----------------------------------------------- | ------ | --------------------------------------------------------------------------------------------------------- |
| `rois_number`                                   | int    | the number of input rois                                                                                  |
| `debug_mode`                                    | bool   | If `true`, subscribe and publish images for visualization.                                                |
| `use_iou`                                       | bool   | calculate IoU both along x-axis and y-axis                                                                |

> | `iou_threshold`                                 | double | the IoU threshold to                                                                                      |

Reviewer 2: how about `min_iou` or `min_iou_threshold`?

Author: updated at [4e4aa5c](https://github.com/autowarefoundation/autoware.universe/pull/2178/commits/4e4aa5c1f2fa2b81ad64bd60bd378b388c75782b).

> | `passthrough_lower_bound_probability_threshold` | double | If `existence_probability` of detected objects is greater than the threshold, they passthrough as outputs |

Reviewer 2: how about `max target existence probability`?

Author: I want to clearfy that do nothing to detected obejcts if the exisitence probability of them is greater than the threshold.

Reviewer 2: ok :+1:

> | `use_roi_probability`                           | float  | use `existence_probability` of ROIs to match with detected objects                                        |

Reviewer 2: ```suggestion
| `use_roi_probability`                           | bool  | use `existence_probability` of ROIs to match with detected objects                                        |
```
